### PR TITLE
Add BatchRaftMessage API

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Add a BatchRaftMessage API

The internal process routine is not changed, just batch some raft messages into one vector, and send the vector in one gRPC streaming chunk instead of send the messages in the streaming directly.

See #3121 .

## What are the type of the changes? (mandatory)
Improve.

## How has this PR been tested? (mandatory)
make dev.

Please describe the tests that you ran to verify your changes. Have you finished unit tests, integration tests, or manual tests? What additional tests would give you greater confidence in this change?

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

**Comment by siddontang** [comment]
how to do the rolling update?

**Comment by hicqu** [comment]
@siddontang If new TiKVs send `BatchRaftMessage`s to old TiKVs, the messages will be rejected by gRPC, and the `Raft` level will retry them.

File: src/server/raft_client.rs [hunk 1]
```diff
@@ -170,9 +172,10 @@ impl RaftClient {
             }
 
             counter += 1;
-            let mut msgs = conn.buffer.take().unwrap();
-            msgs.last_mut().unwrap().1 = WriteFlags::default();
-            if let Err(e) = conn.stream.unbounded_send(msgs) {
+
+            let mut batch_msgs = BatchRaftMessage::new();
```
Comment (siddontang) [inline src/server/raft_client.rs]: do we need to control the total batch size, not too large?

Comment (hicqu) [reply src/server/raft_client.rs]: Here we won't wait any coming messages, so it's not needed?

**Comment by hicqu** [comment]
The fallback machanism is added. PTAL @BusyJay .

**Comment by siddontang** [comment]
have you tested this with sysbench? @hicqu

**Comment by siddontang** [comment]
em, why close it?

**Comment by hicqu** [comment]
According to my test, batch raft messages can't improve too much. Batch messages between TiKV and TiDB can improve much more better. A new PR based on @disksing 's work will be sent after resolving (or prove it's not a problem) a latency issue.

**Comment by hicqu** [comment]
My test is weak. I will test it with 10K small regions again and then post a result.

**Comment by disksing** [comment]
em, I think this PR does not conflict with the gRPC message batch between tidb and tikv. Are you considering using a similar batch method as I have tested for raft messages?

**Pull request closed without merging**